src/main/java/com/mraof/minestuck/item/loot/MSLootEvents.java

```diff
@@ -2,28 +2,144 @@
+			trades.get(2).add((villager, random) -> new MerchantOffer(new ItemStack(Items.EMERALD, 2), new ItemStack(MSItems.STONE_TABLET.get()), 6, 2, 0.05F));
+			trades.get(3).add((villager, random) -> new MerchantOffer(new ItemStack(Items.EMERALD, 5), new ItemStack(MSItems.GREEN_STONE_BRICK_FROG.get()), new ItemStack(MSItems.GREEN_STONE_BRICK_FROG.get(), 2), 12, 7, 0.2F));
+		}
+		
+		if(event.getType() == VillagerProfession.WEAPONSMITH)
+		{
+			trades.get(2).add((villager, random) -> createEnchantedItemOffer(random, new ItemStack(MSItems.LUCERNE_HAMMER.get()), 4, 3, 1, 0.05F));
+		}
+		
+		if(event.getType() == VillagerProfession.FISHERMAN)
+		{
+			trades.get(2).add((villager, random) -> new MerchantOffer(new ItemStack(MSItems.CICADA.get(), 7), new ItemStack(Items.EMERALD, 1), 9, 2, 0.05F));
+			trades.get(2).add((villager, random) -> new MerchantOffer(new ItemStack(MSItems.GRASSHOPPER.get(), 9), new ItemStack(Items.EMERALD, 1), 9, 2, 0.05F));
+		}
+		
+		if(event.getType() == VillagerProfession.LIBRARIAN)
+		{
+			ItemStack[] mediumBookTypes = new ItemStack[]{new ItemStack(MSItems.NONBINARY_CODE.get()), new ItemStack(MSItems.BINARY_CODE.get()), new ItemStack(MSItems.TILLDEATH_HANDBOOK.get()), new ItemStack(MSItems.TABLESTUCK_MANUAL.get()), new ItemStack(MSItems.WISEGUY.get()), new ItemStack(MSItems.SASSACRE_TEXT.get()), new ItemStack(MSItems.FLARP_MANUAL.get())};
+			ItemStack[] overworldBookTypes = new ItemStack[]{new ItemStack(MSItems.COMPLETED_SBURB_CODE.get())};
+			trades.get(2).add((villager, random) -> new MerchantOffer(MSDimensions.isInMedium(villager.getServer(), villager.level.dimension()) ? mediumBookTypes[random.nextInt(mediumBookTypes.length)] : overworldBookTypes[random.nextInt(overworldBookTypes.length)], new ItemStack(Items.EMERALD, 3), 4, 2, 0.05F));
+		}
+	}
+	
+	/**
+	 * Checks for nearby frog temples and creates a map based on an un-generated one if it can be found, seems to cause momentary lag because of this stage at which the map's data is being collected,
+	 * uses TreasureMapForEmeralds in {@link VillagerTrades} as a base
+	 */
+	public static ItemStack createFrogTempleMap(Entity villagerEntity)
+	{
+		Level level = villagerEntity.level;
+		if(level instanceof ServerLevel serverLevel)
+		{
+			BlockPos templePos = serverLevel.findNearestMapFeature(MSTags.Structures.SCANNER_LOCATED, villagerEntity.blockPosition(), 100, true);
+			if(templePos != null)
+			{
+				ItemStack itemstack = MapItem.create(serverLevel, templePos.getX(), templePos.getZ(), (byte) 2, true, true);
+				MapItem.renderBiomePreviewMap(serverLevel, itemstack);
+				MapItemSavedData.addTargetDecoration(itemstack, templePos, "+", MapDecoration.Type.RED_X);
+				itemstack.setHoverName(new TranslatableComponent("filled_map." + MSTags.Structures.SCANNER_LOCATED.toString().toLowerCase(Locale.ROOT)));
+				
+				return itemstack;
+			}
+		}
+		
+		return new ItemStack(MSItems.DICE.get());
```

Comment:
Doesn't seem right. Handle failures like vanilla does instead (ie return a null trade offer if it can't generate one)

Reply:
A null item stack result is not acceptable. You need to return a null merchant offer instead. It may be easier to do if you change the function to return the merchant offer instead. To give a second alternative, you could have it return an `Optional<ItemStack>`, and then you could used it with `createFrogTempleMap().map(mapItem -> ...).orElse(null)`.

Reply:
should be good now

Reply:
Pardon, wrong comment, still need to fix

Reply:
Now returns a merchant offer instead of item stack